### PR TITLE
Switch from `buildSrc/version.properties` to Gradle version catalog (`gradle/libs.versions.toml`) to enable dependabot to perform automated upgrades on common libs

### PR DESCRIPTION
**Pull request by cwperks**
### Description

Opening this PR in draft to explore what it will take to enable dependabot to perform automated upgrades on the dependency versions listed in `buildSrc/version.properties`.

This issue came up for discussion on my very first PR on the project: https://github.com/opensearch-project/OpenSearch/pull/3772. 

Dependabot works on version catalogs: https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#gradle

To test this I pushed to the main branch of my fork and used [Dependabot CLI](https://github.com/dependabot/cli/) to run dependabot in a dry-run mode where it displays what PRs would be created without actually creating a PR.

To test, I created a sample dependabot configuration like this:

```
job:
  package-manager: gradle
  allowed-updates:
    - update-type: all
  source:
    provider: github
    repo: cwperks/opensearch
    directory: /
```

And ran it with `~/go/bin/dependabot update -f ./.github/dependabot_server.yml`

See joda update in the output:

```
updater | +-------------------------------------------------------------------------------------------+
updater | |                            Changes to Dependabot Pull Requests                            |
updater | +---------+---------------------------------------------------------------------------------+
updater | | created | org.apache.ant:ant ( from 1.10.14 to 1.10.15 )                                  |
updater | | created | com.netflix.nebula:gradle-extra-configurations-plugin ( from 10.0.0 to 10.0.1 ) |
updater | | created | com.netflix.nebula:nebula-publishing-plugin ( from 21.0.0 to 21.1.0 )           |
updater | | created | com.netflix.nebula:gradle-info-plugin ( from 12.1.6 to 13.3.0 )                 |
updater | | created | org.apache.rat:apache-rat ( from 0.15 to 0.16.1 )                               |
updater | | created | net.java.dev.jna:jna ( from 5.14.0 to 5.15.0 )                                  |
updater | | created | com.avast.gradle:gradle-docker-compose-plugin ( from 0.17.6 to 0.17.9 )         |
updater | | created | org.apache.maven:maven-model ( from 3.9.6 to 3.9.9 )                            |
updater | | created | com.networknt:json-schema-validator ( from 1.2.0 to 1.5.2 )                     |
updater | | created | org.ajoberstar.grgit:grgit-core ( from 5.2.1 to 5.3.0 )                         |
updater | | created | org.wiremock:wiremock-standalone ( from 3.6.0 to 3.9.1 )                        |
updater | | created | org.spockframework:spock-core ( from 2.3-groovy-3.0 to 2.3-groovy-4.0 )         |
updater | | created | joda-time:joda-time ( from 2.12.7 to 2.13.0 )                                   |
updater | +---------+---------------------------------------------------------------------------------+
```

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/3782

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by cwperks** [comment]
> One of the options of this configuration is to enable auto-merging for a subset of dependencies. I have done so on my own project [here](https://github.com/oshi/oshi/blob/cdf678d60364e3a919c759001ce1ee86db97f3b9/.github/renovate.json#L11-L15).

@peternied Has created something similar to dependabot and opensearch-trigger-bot PRs on the security repo. See the automatic-merged workflow [here](https://github.com/opensearch-project/security/blob/main/.github/workflows/automatic-merges.yml)

**Comment by cwperks** [comment]
There are some external references to the file like [here](https://github.com/opensearch-project/opensearch-build/blob/main/.github/workflows/os-increment-plugin-versions.yml#L76)

**Comment by cwperks** [comment]
Curious to know from maintainers if there is appetite for a change like this, or if status quo with manual updates is sufficient. IMO automated updates help in a number of ways: 1) Less burden on maintainers, 2) better security by ensuring that dependency updates are not missed and 3) using the preferred method of defining common libs in gradle

**Comment by dbwiddis** [comment]
> Curious to know from maintainers if there is appetite for a change like this, or if status quo with manual updates is sufficient. IMO automated updates help in a number of ways: 1) Less burden on maintainers, 2) better security by ensuring that dependency updates are not missed and 3) using the preferred method of defining common libs in gradle

I'd be in favor of automating patch version updates.  Regarding (3) I'd really like to learn more about how plugins consume dpenendencies from a library like this.  It seems we still have to declare our own versions, and worse, we don't get transitive dependencies automatically because they're so tightly controlled.

File: buildSrc/version.properties
Comment (cwperks) [inline buildSrc/version.properties:1]: There are some external references to the `buildSrc/version.properties` file like [here](https://github.com/opensearch-project/opensearch-build/blob/main/.github/workflows/os-increment-plugin-versions.yml#L76)

**Comment by peternied** [comment]
@reta What do you think? You seem to be way more on top of version updates than me, [commit history](https://github.com/opensearch-project/OpenSearch/commits/main/buildSrc/version.properties).

**Comment by reta** [comment]
> @reta What do you think? You seem to be way more on top of version updates than me, [commit history](https://github.com/opensearch-project/OpenSearch/commits/main/buildSrc/version.properties).

Thanks @peternied , it is certainly on my list, will try to get to it shortly (@cwperks sincere apologies)

**Comment by cwperks** [comment]
> @cwperks I think it looks really great, there is only one issue I am concerned about. The versions catalog / TOML file is picked up and updated by dependabot, this is fine. The `version.properties` though (as of today) has to be regenerated manually (or with additional automation) every time TOML file is updated, right?

@reta There is a gradle task in `buildSrc/build.gradle` called `generateVersionProperties` which generates the actual `version.properties` file. At first the `version.properties` file is written to `buildSrc/build`, but after its done generating this file it calls [processResources](https://github.com/opensearch-project/OpenSearch/blob/main/buildSrc/build.gradle#L69-L71) like this:

```
processResources {
  from(generateVersionProperties)
}
```

This will copy the newly created file to `buildSrc/build/resources/main/` which makes it available on the class path to be read in by [VersionProperties](https://github.com/opensearch-project/OpenSearch/blob/main/buildSrc/src/main/java/org/opensearch/gradle/VersionProperties.java#L126-L137)

```
➜  OpenSearch git:(lib-toml) ✗ ls latr buildSrc/build
ls: latr: No such file or directory
buildSrc/build:
classes                  docs                     generated                pluginDescriptors        publications             resources                third-party-audit-config version.properties
distributions            forbidden-apis-config    markers                  pluginUnderTestMetadata  reports                  spotless                 tmp
➜  OpenSearch git:(lib-toml) ✗ ls latr buildSrc/build/resources/main/
ls: latr: No such file or directory
buildSrc/build/resources/main/:
META-INF                     eclipse.settings             fips_java_bcjsse_8.policy    fips_java_sunjsse.security   minimumCompilerVersion       opensearch.properties        version.properties
cacerts.bcfks                fips_java_bcjsse_11.policy   fips_java_bcjsse_8.security  forbidden                    minimumGradleVersion         plugin-descriptor.properties
deb                          fips_java_bcjsse_11.security fips_java_sunjsse.policy     license-headers              minimumRuntimeVersion        test
```

Downstream plugins can still keep their references to versions like [this](https://github.com/opensearch-project/security/blob/main/build.gradle#L475-L484). 

I also added an example of adding `[libraries]` section of the gradle catalog in this PR and showed how it can be used to reference entries in the gradle catalog directly (f.e. `api(libs.jodatime)`). Unfortunately, I don't think we can use the same references in downstream plugins and they would need to keep referencing versions.

**Comment by reta** [comment]
> @reta There is a gradle task in `buildSrc/build.gradle` called `generateVersionProperties` which generates the actual `version.properties` file.

Thanks @cwperks , I totally get this part, I am trying to reconstruct the flow, may be a detailed example would be helpful:
 - the dependabot opens the pull request (it only updates TOML / version catalog, `version.properties` stays unchanged)
 - once dependabot's pull request is merged, the publication of the snapshots happens where the `version.properties` file is regenerated (and published)

Is my understanding correct?

**Comment by cwperks** [comment]
Yes, you are right. It is included in the `build-tools` jar.

For example, the `build-tools` snapshots are located here: https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/gradle/build-tools

**Comment by reta** [comment]
> Yes, you are right. It is included in the `build-tools` jar.

Awesome, thank you. I am wondering if we could completely phase out `versions.properties` and use versions catalog / TOML instead (since it is fenced by `VersionProperties` anyway? (this is just a question, I think we could play safe and always publish the generated version).

**Comment by cwperks** [comment]
> > Yes, you are right. It is included in the `build-tools` jar.
> 
> Awesome, thank you. I am wondering if we could completely phase out `versions.properties` and use versions catalog / TOML instead (since it is fenced by `VersionProperties` anyway)? (this is just a question, I think we could play safe and always publish the generated version).

I would like to see if its possible to consume the new gradle version catalog downstream similar to how the version properties is consumed by plugins.

I like the simplicity:

```
[versions]
joda              = "2.12.7"

[libraries]
jodatime = { group = "joda-time", name = "joda-time", version.ref = "joda" }
```

and in dependencies section of `build.gradle`:

```
dependencies {
    api(libs.jodatime)
}
```

^ With this PR we can do this in any internal module, but not in downstream plugins. For that this PR maintains the publication of the version.properties that is generated from `./gradlew generateVersionProperties`

The gradle version catalog will be really useful in the future because we can also define bundles of dependencies. See https://docs.gradle.org/current/userguide/platforms.html#sub::toml-dependencies-format for what's possible with the version catalog.

**Comment by cwperks** [comment]
To completely phase it out, we need to remove references to the `version.properties` file outside this repo. For instance, I found a reference [here](https://github.com/opensearch-project/opensearch-build/blob/main/.github/workflows/os-increment-plugin-versions.yml#L76)

**Comment by reta** [comment]
> I would like to see if its possible to consume the new gradle version catalog downstream similar to how the version properties is consumed by plugins.

Yes, I think that would be a very logical next step. I am debating if we should be using TOML file or codified `settings.gradle`  (https://github.com/opensearch-project/spring-data-opensearch/blob/main/settings.gradle.kts) but it seems like dependabot does not support versions update for `version` declarations) 

> To completely phase it out, we need to remove references to the `version.properties` file outside this repo. For instance, I found a reference [here](https://github.com/opensearch-project/opensearch-build/blob/main/.github/workflows/os-increment-plugin-versions.yml#L76)

This one references the `buildSrc/version.properties`, not a generated one, right?

**Comment by cwperks** [comment]
> This one references the buildSrc/version.properties, not a generated one, right?

Yep, that's why this PR keeps the file checked in but removes everything other than `opensearch`

**Comment by dbwiddis** [comment]
> The gradle version catalog will be really useful in the future

[Yes.](https://github.com/opensearch-project/opensearch-sdk-java/issues/857#issuecomment-1624334007)

**Comment by reta** [comment]
@cwperks OK, I think I finally glued all pieces together :D (sorry it took so long). So here are 3 parts (or stages) that we have to go through to have dependency management revamped in favour of using versions catalog + be dependabot friendly:
 1. Add `libs.versions.toml` with only `[versions]` section (version catalog), the dependabot would do nothing with that at this stage
 2. Generate `version.properties` out of `libs.versions.toml` (so nothing changes there by and large for consumers)
 3. Add all dependencies to `libs.versions.toml`'s  `[libraries]` section: that would centralize the dependency management and actually would let  dependabot to pick version updates 

Does it make sense? If yes, I would like to ask you remove `[libraries]` section from the TOML file (only keep `[versions]`) so we could minimize any possible (unknown) risks of breaking things. Thanks!

**Comment by cwperks** [comment]
@reta I removed the `[libraries]` section, but keeped the logic in place that parses the `[versions]` section from the TOML file to generate the `version.properties` file. That way if we add additional sections in the toml file in the future, the logic to handle it is already in place.

**Comment by reta** [comment]
> @reta I removed the `[libraries]` section, but keeped the logic in place that parses the `[versions]` section from the TOML file to generate the `version.properties` file. That way if we add additional sections in the toml file in the future, the logic to handle it is already in place.

@cwperks about that, do you mind if I push a small change to this particular part? I prefer us to not do manual parsing. Thank you.

**Comment by cwperks** [comment]
> > @reta I removed the `[libraries]` section, but keeped the logic in place that parses the `[versions]` section from the TOML file to generate the `version.properties` file. That way if we add additional sections in the toml file in the future, the logic to handle it is already in place.
> 
> @cwperks about that, do you mind if I push a small change to this particular part? I prefer us to not do manual parsing. Thank you.

Go ahead! You can push to any of my PRs ;)

**Comment by reta** [comment]
@dblock @dbwiddis @andrross @peternied I think this is first step in right direction (to replace ad-hoc version management with Gradle version catalogs, see please larger picture https://github.com/opensearch-project/OpenSearch/pull/16284#issuecomment-2422497893). Curious what do you think folks, any concerns?  LGTM to me, thanks a lot @cwperks !

**Review by reta — APPROVED** [review]

**Comment by dbwiddis** [comment]
> Curious what do you think folks, any concerns?

How are transitive dependencies handled?   I've experienced frequent issues in plugins where we need Dependency X which transitively depends on Y.  OpenSearch has Y at an earlier version but doesn't have X at all.  Will we still be able to force a higher version?

Will plugins be able to access the version number that OpenSearch has easily (version catalog) so that we don't have to maintain (synchronized) version bumps ourselves?

**Comment by reta** [comment]
Thanks @dbwiddis 

> How are transitive dependencies handled?

If I am not mistaken, this is out of scope of version catalogs BUT is in scope of the individual modules: it all depends on how the module import the dependency (runtime, compile, ...)

>  Will we still be able to force a higher version?

I believe the `force` mechanism will stay explicit (I could be mistaken)

> Will plugins be able to access the version number that OpenSearch has easily (version catalog) so that we don't have to maintain (synchronized) version bumps ourselves?

Yes! That's the end goal!

**Comment by dbwiddis** [comment]
> Yes! That's the end goal!
🎉

**Comment by cwperks** [comment]
Sync'ed with main to resolve conflicts and ensured the latest versions on main for version.properties are copied to libs.versions.toml

**Comment by cwperks** [comment]
@reta Is this same error appearing on main?

```
Caused by:
        org.mockito.exceptions.****.MockitoException: Could not modify all classes [class java.lang.Object, interface java.io.Closeable, interface org.apache.hc.client5.http.async.HttpAsyncClient, class org.apache.hc.client5.http.impl.async.CloseableHttpAsyncClient, interface java.lang.AutoCloseable, interface org.apache.hc.core5.io.ModalCloseable]
            at app//net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:168)
            at app//net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:399)
            at app//net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:190)
            at app//net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:410)
            ... 2 more

            Caused by:
            java.lang.IllegalStateException: 
            Byte Buddy could not instrument all classes within the mock's type hierarchy

            This problem should never occur for javac-compiled classes. This problem has been observed for classes that are:
             - Compiled by older versions of scalac
             - Classes that are part of the Android distribution
                at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.triggerRetransformation(InlineBytecodeGenerator.java:285)
                at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.mockClass(InlineBytecodeGenerator.java:218)
                at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator.lambda$mockClass$0(TypeCachingBytecodeGenerator.java:78)
                at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:168)
                at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:399)
                at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:190)
                at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:410)
                at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator.mockClass(TypeCachingBytecodeGenerator.java:75)
                at org.mockito.internal.creation.bytebuddy.InlineDelegateByteBuddyMockMaker.createMockType(InlineDelegateByteBuddyMockMaker.java:414)
                at org.mockito.internal.creation.bytebuddy.InlineDelegateByteBuddyMockMaker.doCreateMock(InlineDelegateByteBuddyMockMaker.java:373)
                at org.mockito.internal.creation.bytebuddy.InlineDelegateByteBuddyMockMaker.createMock(InlineDelegateByteBuddyMockMaker.java:352)
                at org.mockito.internal.creation.bytebuddy.InlineByteBuddyMockMaker.createMock(InlineByteBuddyMockMaker.java:56)
                at org.mockito.internal.util.MockUtil.createMock(MockUtil.java:99)
                at org.mockito.internal.MockitoCore.mock(MockitoCore.java:84)
                at org.mockito.Mockito.mock(Mockito.java:2104)
                at org.mockito.Mockito.mock(Mockito.java:2019)
                ... 2 more

                Caused by:
                java.lang.IllegalArgumentException: Java 23 (67) is not supported by the current version of Byte Buddy which officially supports Java 22 (66) - update Byte Buddy or set net.bytebuddy.experimental as a VM property
                    at net.bytebuddy.utility.OpenedClassReader.of(OpenedClassReader.java:100)
                    at net.bytebuddy.dynamic.scaffold.TypeWriter$Default$ForInlining.create(TypeWriter.java:4011)
                    at net.bytebuddy.dynamic.scaffold.TypeWriter$Default.make(TypeWriter.java:2224)
                    at net.bytebuddy.dynamic.DynamicType$Builder$AbstractBase$UsingTypeWriter.make(DynamicType.java:4055)
                    at net.bytebuddy.dynamic.DynamicType$Builder$AbstractBase.make(DynamicType.java:3739)
                    at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.transform(InlineBytecodeGenerator.java:402)
                    at java.instrument/java.lang.instrument.ClassFileTransformer.transform(ClassFileTransformer.java:242)
                    at java.instrument/sun.instrument.TransformerManager.transform(TransformerManager.java:188)
                    at java.instrument/sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:610)
                    at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses0(Native Method)
                    at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses(InstrumentationImpl.java:225)
                    at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.triggerRetransformation(InlineBytecodeGenerator.java:281)
                    ... 17 more

> Task :build-tools:compileIntegTestGroovy
> Task :build-tools:processIntegTestResources
> Task :build-tools:integTestClasses

> Task :client:rest:test

Tests with failures:
 - org.opensearch.client.RestClientTests.testIsRunning
 - org.opensearch.client.RestClientTests.testStreamWithUnsupportedMethod
 - org.opensearch.client.RestClientTests.testPerformAsyncWithUnsupportedMethod
 - org.opensearch.client.RestClientTests.testCloseIsIdempotent
 - org.opensearch.client.RestClientTests.testSetNodesDuplicatedHosts
 - org.opensearch.client.RestClientTests.testSetNodesWrongArguments
 - org.opensearch.client.RestClientTests.testSetNodesPreservesOrdering
 - org.opensearch.client.RestClientTests.testPerformAsyncWithWrongEndpoint

97 tests completed, 8 failed, 1 skipped

> Task :client:rest:test FAILED
```

**Comment by reta** [comment]
> @reta Is this same error appearing on main?

@cwperks no, sadly version.properties and TOML file diverged :( needs to be reconciled

File: gradle/libs.versions.toml
Comment (reta) [reply gradle/libs.versions.toml:66]: May be more versions need to be reconciled ..

**Comment by cwperks** [comment]
My bad, I did not update the bytebuddy or mockito versions to the latest on main. Its fully reconciled now.

**Review by dblock — APPROVED** [review]

**Review by dblock — APPROVED** [review]

**Review by andrross — APPROVED** [review]

**Comment by reta** [comment]
@cwperks apologies, mind please taking `2.x` backport? thank you!

**Comment by cwperks** [comment]
Yes, I am looking into in now. 1 min

**Comment by cwperks** [comment]
@reta opened a backport here: https://github.com/opensearch-project/OpenSearch/pull/16508

Thank you to everyone that helped move this forward :)

**Pull request merged**
